File: sdk/src/main/java/com/libreshockwave/lingo/Opcode.java
package com.libreshockwave.lingo;

import java.util.HashMap;
import java.util.Map;

/**
 * Lingo bytecode opcodes.
 * Single-byte opcodes (0x00-0x3F) take no arguments.
 * Multi-byte opcodes (0x40+) have variable-length arguments.
 */
public enum Opcode {

    // Single-byte opcodes (no arguments)
    INVALID(0x00, "invalid", 0),
    RET(0x01, "ret", 0),
    RET_FACTORY(0x02, "retFactory", 0),
    PUSH_ZERO(0x03, "pushZero", 0),
    MUL(0x04, "mul", 0),
    ADD(0x05, "add", 0),
    SUB(0x06, "sub", 0),
    DIV(0x07, "div", 0),
    MOD(0x08, "mod", 0),
    INV(0x09, "inv", 0),
    JOIN_STR(0x0A, "joinStr", 0),
    JOIN_PAD_STR(0x0B, "joinPadStr", 0),
    LT(0x0C, "lt", 0),
    LT_EQ(0x0D, "ltEq", 0),
    NT_EQ(0x0E, "ntEq", 0),
    EQ(0x0F, "eq", 0),
    GT(0x10, "gt", 0),
    GT_EQ(0x11, "gtEq", 0),
    AND(0x12, "and", 0),
    OR(0x13, "or", 0),
    NOT(0x14, "not", 0),
    CONTAINS_STR(0x15, "containsStr", 0),
    CONTAINS_0_STR(0x16, "contains0Str", 0),
    GET_CHUNK(0x17, "getChunk", 0),
    HILITE_CHUNK(0x18, "hiliteChunk", 0),
    ONTO_SPR(0x19, "ontoSpr", 0),
    INTO_SPR(0x1A, "intoSpr", 0),
    GET_FIELD(0x1B, "getField", 0),
    START_TELL(0x1C, "startTell", 0),
    END_TELL(0x1D, "endTell", 0),
    PUSH_LIST(0x1E, "pushList", 0),
    PUSH_PROP_LIST(0x1F, "pushPropList", 0),
    SWAP(0x21, "swap", 0),
    CALL_JAVASCRIPT(0x26, "callJavaScript", 0),

    // Multi-byte opcodes (with arguments)
    PUSH_INT8(0x41, "pushInt8", 1),
    PUSH_ARG_LIST_NO_RET(0x42, "pushArgListNoRet", 1),
    PUSH_ARG_LIST(0x43, "pushArgList", 1),
    PUSH_CONS(0x44, "pushCons", 1),
    PUSH_SYMB(0x45, "pushSymb", 1),
    PUSH_VAR_REF(0x46, "pushVarRef", 1),
    GET_GLOBAL2(0x48, "getGlobal2", 1),
    GET_GLOBAL(0x49, "getGlobal", 1),
    GET_PROP(0x4A, "getProp", 1),
    GET_PARAM(0x4B, "getParam", 1),
    GET_LOCAL(0x4C, "getLocal", 1),
    SET_GLOBAL2(0x4E, "setGlobal2", 1),
    SET_GLOBAL(0x4F, "setGlobal", 1),
    SET_PROP(0x50, "setProp", 1),
    SET_PARAM(0x51, "setParam", 1),
    SET_LOCAL(0x52, "setLocal", 1),
    JMP(0x53, "jmp", 1),
    END_REPEAT(0x54, "endRepeat", 1),
    JMP_IF_Z(0x55, "jmpIfZ", 1),
    LOCAL_CALL(0x56, "localCall", 1),
    EXT_CALL(0x57, "extCall", 1),
    OBJ_CALL_V4(0x58, "objCallV4", 1),
    PUT(0x59, "put", 1),
    PUT_CHUNK(0x5A, "putChunk", 1),
    DELETE_CHUNK(0x5B, "deleteChunk", 1),
    GET(0x5C, "get", 1),
    SET(0x5D, "set", 1),
    GET_MOVIE_PROP(0x5F, "getMovieProp", 1),
    SET_MOVIE_PROP(0x60, "setMovieProp", 1),
    GET_OBJ_PROP(0x61, "getObjProp", 1),
    SET_OBJ_PROP(0x62, "setObjProp", 1),
    TELL_CALL(0x63, "tellCall", 1),
    PEEK(0x64, "peek", 1),
    POP(0x65, "pop", 1),
    THE_BUILTIN(0x66, "theBuiltin", 1),
    OBJ_CALL(0x67, "objCall", 1),
    PUSH_CHUNK_VAR_REF(0x6D, "pushChunkVarRef", 1),
    PUSH_INT16(0x6E, "pushInt16", 2),
    PUSH_INT32(0x6F, "pushInt32", 4),
    GET_CHAINED_PROP(0x70, "getChainedProp", 1),
    PUSH_FLOAT32(0x71, "pushFloat32", 4),
    GET_TOP_LEVEL_PROP(0x72, "getTopLevelProp", 1),
    NEW_OBJ(0x73, "newObj", 1);

    private static final Map<Integer, Opcode> BY_CODE = new HashMap<>();

    static {
        for (Opcode op : values()) {
            BY_CODE.put(op.code, op);
        }
    }

    private final int code;
    private final String mnemonic;
    private final int argBytes;

    Opcode(int code, String mnemonic, int argBytes) {
        this.code = code;
        this.mnemonic = mnemonic;
        this.argBytes = argBytes;
    }

    public int getCode() {
        return code;
    }

    public String getMnemonic() {
        return mnemonic;
    }

    public int getArgBytes() {
        return argBytes;
    }

    public boolean isSingleByte() {
        return code < 0x40;
    }

    public boolean isMultiByte() {
        return code >= 0x40;
    }

    public boolean isJump() {
        return this == JMP || this == JMP_IF_Z || this == END_REPEAT;
    }

    public boolean isCall() {
        return this == LOCAL_CALL || this == EXT_CALL || this == OBJ_CALL || this == OBJ_CALL_V4 || this == TELL_CALL;
    }

    public boolean isReturn() {
        return this == RET || this == RET_FACTORY;
    }

    public boolean isPush() {
        return switch (this) {
            case PUSH_ZERO, PUSH_INT8, PUSH_INT16, PUSH_INT32, PUSH_FLOAT32,
                 PUSH_CONS, PUSH_SYMB, PUSH_VAR_REF, PUSH_CHUNK_VAR_REF,
                 PUSH_LIST, PUSH_PROP_LIST, PUSH_ARG_LIST, PUSH_ARG_LIST_NO_RET -> true;
            default -> false;
        };
    }

    public static Opcode fromCode(int code) {
        Opcode op = BY_CODE.get(code);
        if (op == null) {
            return INVALID;
        }
        return op;
    }

    @Override
    public String toString() {
        return mnemonic;
    }
}

File: cpp/src/lingo/Opcode.cpp
#include "libreshockwave/lingo/Opcode.hpp"

namespace libreshockwave::lingo {

int code(Opcode opcode) {
    return static_cast<int>(opcode);
}

std::string_view mnemonic(Opcode opcode) {
    switch (opcode) {
        case Opcode::INVALID: return "invalid";
        case Opcode::RET: return "ret";
        case Opcode::RET_FACTORY: return "retFactory";
        case Opcode::PUSH_ZERO: return "pushZero";
        case Opcode::MUL: return "mul";
        case Opcode::ADD: return "add";
        case Opcode::SUB: return "sub";
        case Opcode::DIV: return "div";
        case Opcode::MOD: return "mod";
        case Opcode::INV: return "inv";
        case Opcode::JOIN_STR: return "joinStr";
        case Opcode::JOIN_PAD_STR: return "joinPadStr";
        case Opcode::LT: return "lt";
        case Opcode::LT_EQ: return "ltEq";
        case Opcode::NT_EQ: return "ntEq";
        case Opcode::EQ: return "eq";
        case Opcode::GT: return "gt";
        case Opcode::GT_EQ: return "gtEq";
        case Opcode::AND: return "and";
        case Opcode::OR: return "or";
        case Opcode::NOT: return "not";
        case Opcode::CONTAINS_STR: return "containsStr";
        case Opcode::CONTAINS_0_STR: return "contains0Str";
        case Opcode::GET_CHUNK: return "getChunk";
        case Opcode::HILITE_CHUNK: return "hiliteChunk";
        case Opcode::ONTO_SPR: return "ontoSpr";
        case Opcode::INTO_SPR: return "intoSpr";
        case Opcode::GET_FIELD: return "getField";
        case Opcode::START_TELL: return "startTell";
        case Opcode::END_TELL: return "endTell";
        case Opcode::PUSH_LIST: return "pushList";
        case Opcode::PUSH_PROP_LIST: return "pushPropList";
        case Opcode::SWAP: return "swap";
        case Opcode::CALL_JAVASCRIPT: return "callJavaScript";
        case Opcode::PUSH_INT8: return "pushInt8";
        case Opcode::PUSH_ARG_LIST_NO_RET: return "pushArgListNoRet";
        case Opcode::PUSH_ARG_LIST: return "pushArgList";
        case Opcode::PUSH_CONS: return "pushCons";
        case Opcode::PUSH_SYMB: return "pushSymb";
        case Opcode::PUSH_VAR_REF: return "pushVarRef";
        case Opcode::GET_GLOBAL2: return "getGlobal2";
        case Opcode::GET_GLOBAL: return "getGlobal";
        case Opcode::GET_PROP: return "getProp";
        case Opcode::GET_PARAM: return "getParam";
        case Opcode::GET_LOCAL: return "getLocal";
        case Opcode::SET_GLOBAL2: return "setGlobal2";
        case Opcode::SET_GLOBAL: return "setGlobal";
        case Opcode::SET_PROP: return "setProp";
        case Opcode::SET_PARAM: return "setParam";
        case Opcode::SET_LOCAL: return "setLocal";
        case Opcode::JMP: return "jmp";
        case Opcode::END_REPEAT: return "endRepeat";
        case Opcode::JMP_IF_Z: return "jmpIfZ";
        case Opcode::LOCAL_CALL: return "localCall";
        case Opcode::EXT_CALL: return "extCall";
        case Opcode::OBJ_CALL_V4: return "objCallV4";
        case Opcode::PUT: return "put";
        case Opcode::PUT_CHUNK: return "putChunk";
        case Opcode::DELETE_CHUNK: return "deleteChunk";
        case Opcode::GET: return "get";
        case Opcode::SET: return "set";
        case Opcode::GET_MOVIE_PROP: return "getMovieProp";
        case Opcode::SET_MOVIE_PROP: return "setMovieProp";
        case Opcode::GET_OBJ_PROP: return "getObjProp";
        case Opcode::SET_OBJ_PROP: return "setObjProp";
        case Opcode::TELL_CALL: return "tellCall";
        case Opcode::PEEK: return "peek";
        case Opcode::POP: return "pop";
        case Opcode::THE_BUILTIN: return "theBuiltin";
        case Opcode::OBJ_CALL: return "objCall";
        case Opcode::PUSH_CHUNK_VAR_REF: return "pushChunkVarRef";
        case Opcode::PUSH_INT16: return "pushInt16";
        case Opcode::PUSH_INT32: return "pushInt32";
        case Opcode::GET_CHAINED_PROP: return "getChainedProp";
        case Opcode::PUSH_FLOAT32: return "pushFloat32";
        case Opcode::GET_TOP_LEVEL_PROP: return "getTopLevelProp";
        case Opcode::NEW_OBJ: return "newObj";
    }
    return "invalid";
}

int argBytes(Opcode opcode) {
    switch (opcode) {
        case Opcode::PUSH_INT16:
            return 2;
        case Opcode::PUSH_INT32:
        case Opcode::PUSH_FLOAT32:
            return 4;
        case Opcode::PUSH_INT8:
        case Opcode::PUSH_ARG_LIST_NO_RET:
        case Opcode::PUSH_ARG_LIST:
        case Opcode::PUSH_CONS:
        case Opcode::PUSH_SYMB:
        case Opcode::PUSH_VAR_REF:
        case Opcode::GET_GLOBAL2:
        case Opcode::GET_GLOBAL:
        case Opcode::GET_PROP:
        case Opcode::GET_PARAM:
        case Opcode::GET_LOCAL:
        case Opcode::SET_GLOBAL2:
        case Opcode::SET_GLOBAL:
        case Opcode::SET_PROP:
        case Opcode::SET_PARAM:
        case Opcode::SET_LOCAL:
        case Opcode::JMP:
        case Opcode::END_REPEAT:
        case Opcode::JMP_IF_Z:
        case Opcode::LOCAL_CALL:
        case Opcode::EXT_CALL:
        case Opcode::OBJ_CALL_V4:
        case Opcode::PUT:
        case Opcode::PUT_CHUNK:
        case Opcode::DELETE_CHUNK:
        case Opcode::GET:
        case Opcode::SET:
        case Opcode::GET_MOVIE_PROP:
        case Opcode::SET_MOVIE_PROP:
        case Opcode::GET_OBJ_PROP:
        case Opcode::SET_OBJ_PROP:
        case Opcode::TELL_CALL:
        case Opcode::PEEK:
        case Opcode::POP:
        case Opcode::THE_BUILTIN:
        case Opcode::OBJ_CALL:
        case Opcode::PUSH_CHUNK_VAR_REF:
        case Opcode::GET_CHAINED_PROP:
        case Opcode::GET_TOP_LEVEL_PROP:
        case Opcode::NEW_OBJ:
            return 1;
        default:
            return 0;
    }
}

Opcode opcodeFromCode(int value) {
    switch (value) {
        case 0x00: return Opcode::INVALID;
        case 0x01: return Opcode::RET;
        case 0x02: return Opcode::RET_FACTORY;
        case 0x03: return Opcode::PUSH_ZERO;
        case 0x04: return Opcode::MUL;
        case 0x05: return Opcode::ADD;
        case 0x06: return Opcode::SUB;
        case 0x07: return Opcode::DIV;
        case 0x08: return Opcode::MOD;
        case 0x09: return Opcode::INV;
        case 0x0A: return Opcode::JOIN_STR;
        case 0x0B: return Opcode::JOIN_PAD_STR;
        case 0x0C: return Opcode::LT;
        case 0x0D: return Opcode::LT_EQ;
        case 0x0E: return Opcode::NT_EQ;
        case 0x0F: return Opcode::EQ;
        case 0x10: return Opcode::GT;
        case 0x11: return Opcode::GT_EQ;
        case 0x12: return Opcode::AND;
        case 0x13: return Opcode::OR;
        case 0x14: return Opcode::NOT;
        case 0x15: return Opcode::CONTAINS_STR;
        case 0x16: return Opcode::CONTAINS_0_STR;
        case 0x17: return Opcode::GET_CHUNK;
        case 0x18: return Opcode::HILITE_CHUNK;
        case 0x19: return Opcode::ONTO_SPR;
        case 0x1A: return Opcode::INTO_SPR;
        case 0x1B: return Opcode::GET_FIELD;
        case 0x1C: return Opcode::START_TELL;
        case 0x1D: return Opcode::END_TELL;
        case 0x1E: return Opcode::PUSH_LIST;
        case 0x1F: return Opcode::PUSH_PROP_LIST;
        case 0x21: return Opcode::SWAP;
        case 0x26: return Opcode::CALL_JAVASCRIPT;
        case 0x41: return Opcode::PUSH_INT8;
        case 0x42: return Opcode::PUSH_ARG_LIST_NO_RET;
        case 0x43: return Opcode::PUSH_ARG_LIST;
        case 0x44: return Opcode::PUSH_CONS;
        case 0x45: return Opcode::PUSH_SYMB;
        case 0x46: return Opcode::PUSH_VAR_REF;
        case 0x48: return Opcode::GET_GLOBAL2;
        case 0x49: return Opcode::GET_GLOBAL;
        case 0x4A: return Opcode::GET_PROP;
        case 0x4B: return Opcode::GET_PARAM;
        case 0x4C: return Opcode::GET_LOCAL;
        case 0x4E: return Opcode::SET_GLOBAL2;
        case 0x4F: return Opcode::SET_GLOBAL;
        case 0x50: return Opcode::SET_PROP;
        case 0x51: return Opcode::SET_PARAM;
        case 0x52: return Opcode::SET_LOCAL;
        case 0x53: return Opcode::JMP;
        case 0x54: return Opcode::END_REPEAT;
        case 0x55: return Opcode::JMP_IF_Z;
        case 0x56: return Opcode::LOCAL_CALL;
        case 0x57: return Opcode::EXT_CALL;
        case 0x58: return Opcode::OBJ_CALL_V4;
        case 0x59: return Opcode::PUT;
        case 0x5A: return Opcode::PUT_CHUNK;
        case 0x5B: return Opcode::DELETE_CHUNK;
        case 0x5C: return Opcode::GET;
        case 0x5D: return Opcode::SET;
        case 0x5F: return Opcode::GET_MOVIE_PROP;
        case 0x60: return Opcode::SET_MOVIE_PROP;
        case 0x61: return Opcode::GET_OBJ_PROP;
        case 0x62: return Opcode::SET_OBJ_PROP;
        case 0x63: return Opcode::TELL_CALL;
        case 0x64: return Opcode::PEEK;
        case 0x65: return Opcode::POP;
        case 0x66: return Opcode::THE_BUILTIN;
        case 0x67: return Opcode::OBJ_CALL;
        case 0x6D: return Opcode::PUSH_CHUNK_VAR_REF;
        case 0x6E: return Opcode::PUSH_INT16;
        case 0x6F: return Opcode::PUSH_INT32;
        case 0x70: return Opcode::GET_CHAINED_PROP;
        case 0x71: return Opcode::PUSH_FLOAT32;
        case 0x72: return Opcode::GET_TOP_LEVEL_PROP;
        case 0x73: return Opcode::NEW_OBJ;
        default: return Opcode::INVALID;
    }
}

bool isSingleByte(Opcode opcode) {
    return code(opcode) < 0x40;
}

bool isMultiByte(Opcode opcode) {
    return code(opcode) >= 0x40;
}

bool isJump(Opcode opcode) {
    return opcode == Opcode::JMP || opcode == Opcode::JMP_IF_Z || opcode == Opcode::END_REPEAT;
}

bool isCall(Opcode opcode) {
    return opcode == Opcode::LOCAL_CALL ||
           opcode == Opcode::EXT_CALL ||
           opcode == Opcode::OBJ_CALL ||
           opcode == Opcode::OBJ_CALL_V4 ||
           opcode == Opcode::TELL_CALL;
}

bool isReturn(Opcode opcode) {
    return opcode == Opcode::RET || opcode == Opcode::RET_FACTORY;
}

bool isPush(Opcode opcode) {
    switch (opcode) {
        case Opcode::PUSH_ZERO:
        case Opcode::PUSH_INT8:
        case Opcode::PUSH_INT16:
        case Opcode::PUSH_INT32:
        case Opcode::PUSH_FLOAT32:
        case Opcode::PUSH_CONS:
        case Opcode::PUSH_SYMB:
        case Opcode::PUSH_VAR_REF:
        case Opcode::PUSH_CHUNK_VAR_REF:
        case Opcode::PUSH_LIST:
        case Opcode::PUSH_PROP_LIST:
        case Opcode::PUSH_ARG_LIST:
        case Opcode::PUSH_ARG_LIST_NO_RET:
            return true;
        default:
            return false;
    }
}

} // namespace libreshockwave::lingo
